realm/realm-library/src/objectServer/java/io/realm/SyncSession.java

```diff
@@ -118,6 +124,63 @@ void close() {
         clearScheduledAccessTokenRefresh();
     }
 
+    // This method will be called once all changes have been downloaded.
+    // This method might be called on another thread than the one that called `downloadAllServerChanges`.
+    // Be very careful with synchronized blocks.
+    // If the native listener was successfully registered, Object Store guarantees that this method will be called at
+    // least once, even if the session is closed.
+    void notifyAllChangesDownloaded(Long errorcode, String errorMessage) {
+        if (waitingForServerChanges != null) {
+            waitingForServerChanges.handleResult(errorcode, errorMessage);
+        }
+    }
+
+    /**
+     * Calling this method will block until all known remote changes have been downloaded and applied to the Realm.
+     * This will involve network access, so calling this method should only be done from a non-UI thread.
+     * <p>
+     * If the device is offline, this method might never return.
+     * <p>
+     * This method cannot be called before the session has been started.
+     *
+     * @throws IllegalStateException if called on the Android main thread.
+     */
+    public void downloadAllServerChanges() {
+        checkMainThread("downloadAllServerChanges() cannot be called from the main thread.");
+
+        // Blocking only happens at the Java layer. To prevent deadlocking the underlying SyncSession we register
+        // an async listener there and let it callback to the Java Session when done. This feels icky at best, but
+        // since all operations on the SyncSession operate under a shared mutex, we would prevent all other actions on the
+        // session, including trying to stop it.
+        // In Java we cannot lock on the Session object either since it will prevent any attempt at modifying the
+        // lifecycle while it is in a waiting state. Thus we use a specialised mutex.
+        synchronized (waitForChangesMutex) {
+            if (!isClosed) {
+                waitingForServerChanges = new WaitForServerChangesWrapper();
+                boolean listenerRegistered = nativeWaitForDownloadCompletion(configuration.getPath());
+                if (!listenerRegistered) {
+                    waitingForServerChanges = null;
+                    throw new ObjectServerError(ErrorCode.UNKNOWN, "It was not possible to download all changes. Has the SyncClient been started?");
+                }
+                waitingForServerChanges.waitForServerChanges();
+
+                // This might return after the session was closed. In that case, just ignore any result
+                if (!isClosed) {
+                    if (!waitingForServerChanges.isSuccess()) {
+                        waitingForServerChanges.throwExceptionIfNeeded();
```

Comment:
`SyncSession` needs to be completely thread-safe, and `notifyAllChangesDownloaded` uses the presence of this class as an indicator. This probably wasn't thread safe either though, so I converted it to an AtomicReference instead.